### ISSUE TITLE
False positive: InvalidPropertyAssignmentValue

### ISSUE THREAD

**Issue by rzvc** [vimeo/psalm#8002]
```php
/** @template T */
class Foo
{
}

class Bar
{
    /** @var ?Foo<array{a?: string}> */
    public $test = null;
}

$bar = new Bar();

/** @var Foo<array{a: string}> $x */
$bar->test = $x;
```
https://psalm.dev/r/c4c3d22b69
It appears it won't accept non-optional properties in the place of optional ones. Might have something to do with the templates because this works: https://psalm.dev/r/30fb22a8fa

**Edit:** Appears to exhibit the same behaviour with unions too: https://psalm.dev/r/2e196f1da4

**Edit 2:** I narrowed it down to `GenericTypeComparator::isContainedBy()`, line ~185. It appears to be done on purpose, and I understand why. Now the question is how do I deal with it. My actual `Foo`, has a method that returns T, and none that take T, so it's an output only situation, where this restriction is not necessary.

Suggestions?

**Edit 3:** I have a suggestion. Instead of issuing an `InvalidPropertyAssignmentValue`, it should issue something more specific, that shows that the types is contained, but not identical. That can then be suppress and `InvalidPropertyAssignmentValue` would still get through, which would indicate an actual error in my case.

**Comment by AndrolGenhald** [comment]
This is problematic in general because of the [barking cat](https://psalm.dev/docs/annotating_code/templated_annotations/#template-covariance) problem: https://psalm.dev/r/d85b337cc7 https://3v4l.org/lRXf0

If your code allows it, you can declare the template covariant: https://psalm.dev/r/37d9f6251e

**Comment by rzvc** [comment]
Marking it covariant worked. Thanks.

Yes, it would have helped if it was listed somewhere.
